benches/dapol.rs

```diff
@@ -1,175 +1,385 @@
+
+                if u > max_users_for_height {
+                    break;
+                }
+
+                let tup: (Height, MaxThreadCount, u64) =
+                    (Height::from(h), MaxThreadCount::from(*t), u);
 
-    // this benchmark depends on the tree height and not the number of leaves,
-    // so we just pick the smallest number of leaves
-    let num_leaves = NUM_LEAVES[0];
-    for &tree_height in TREE_HEIGHTS.iter() {
-        let items = build_item_list(num_leaves, tree_height);
-        let mut rng = thread_rng();
-        let item_range = Uniform::new(0usize, num_leaves);
-
-        let dapol = build_dapol_tree::<blake3::Hasher, RangeProofSplitting>(&items, tree_height);
-        group.bench_function(BenchmarkId::new("splitting", tree_height), |bench| {
-            bench.iter_batched(
-                || {
-                    // generate a proof
-                    let item_idx = rng.sample(item_range);
-                    let tree_index = &items[item_idx].0;
-                    (item_idx, dapol.generate_proof(tree_index).unwrap())
-                },
-                |(item_idx, proof)| {
-                    // time proof verification
-                    proof.verify(&dapol.root(), &items[item_idx].1.get_proof_node())
-                },
-                BatchSize::SmallInput,
-            );
-        });
-
-        let dapol = build_dapol_tree::<blake3::Hasher, RangeProofPadding>(&items, tree_height);
-        group.bench_function(BenchmarkId::new("padding", tree_height), |bench| {
-            bench.iter_batched(
-                || {
-                    // generate a proof
-                    let item_idx = rng.sample(item_range);
-                    let tree_index = &items[item_idx].0;
-                    (item_idx, dapol.generate_proof(tree_index).unwrap())
-                },
-                |(item_idx, proof)| {
-                    // time proof verification
-                    proof.verify(&dapol.root(), &items[item_idx].1.get_proof_node())
-                },
-                BatchSize::SmallInput,
-            );
-        });
+                group.bench_function(
+                    BenchmarkId::new(
+                        "build_tree",
+                        format!("{:?}/{:?}/NUM_USERS: {:?}", &tup.0, &tup.1, &tup.2),
+                    ),
+                    |bench| {
+                        bench.iter(|| {
+                            ndm_smt = setup::build_ndm_smt(tup.clone()).ok();
+                        });
+                    },
+                );
+
+                if let Some(inner) = &ndm_smt {
+                    setup::serialize_tree(inner, PathBuf::from("./target"))
+                } else {
+                    return Err("Tree not found");
+                }
+            }
+
+            // prevent duplicate max thread counts
+            if *t == max_thread_count {
+                break;
+            }
```

Comment:
When does this situation arise?

Reply:
When the `max_thread_count` vector contains two values that are both the max value. E.g., when I run the benchmark on a tree height of 16, I get a panic because 16 appears twice, due to how the vector is constructed. 

I.e., `max_thread_counts: [16, 20, 24, 28, 16]`

Reply:
No worries. This seems to have sorted it, thanks